src/librustc_middle/ty/print/pretty.rs

```diff
@@ -1159,6 +1146,66 @@ pub trait PrettyPrinter<'tcx>:
                 p!(pretty_print_byte_str(byte_str));
                 Ok(self)
             }
+
+            // Aggregates, printed as array/tuple/struct/variant construction syntax.
+            //
+            // NB: the `has_param_types_or_consts` check ensures that we can use
+            // the `destructure_const` query with an empty `ty::ParamEnv` without
+            // introducing ICEs (e.g. via `layout_of`) from missing bounds.
+            // E.g. `transmute([0usize; 2]): (u8, *mut T)` needs to know `T: Sized`
```

Comment:
can this happen at all? Const evaluation must already have had to know the type in order to produce a `ConstValue`. While the *type* may still be generic, in case of polymorphic const eval working out because of e.g. `PhantomData<T>` being a known ZST irrelevant of the type, any `layout_of` calls we do during `destructure_const` should also figure this out.

Reply:
`*mut T` has a known layout irrespective of `T` *as long as* `T: Sized` is in the `ParamEnv`, that's why I used it as an example. You can use `size_of::<*mut T>()` as an explicit discriminant of an `enum`, for example.

Reply:
right, but here we already have a `ConstValue`, which must have gone through some monomorphic evaluation at some point.

Reply:
Nope, there is *nothing* requiring monomorphic const-eval, and I have [this example](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=c4866680fd249d536b6068903dd36cec) to prove it (I made that to prove a point in #70453).

Reply:
I guess your point is that the type can't have type parameters? Sorry, I was getting turned around. `const` generics/array lengths shouldn't hit this.

So this could only be a problem with printing MIR after constant-folding something that uses type parameters but doesn't depend on them layout-wise.

Riiight, which is why I was thinking MIR printing could create a `FmtPrinter` with a `ParamEnv`.